Lab_Framework. Загрузка

UlyanaRomanova opened the pull request.

UlyanaRomanova committed:
Lab_Framework. Загрузка

DuskTraum commented:
-IsElementPresent лучше перенести в страницы и использовать его оттуда, тесты не должны работать с элементами
-GetAttributeFields() лучше пусть возвращает  объект типа Product, а не List. Тогда его использовать можно будет непосредственно в ассерте
- OpenProduct для сравнения в ассертах используй свой продукт productCoffee, а то получается дублирование значений
- Не стоит возвращать с сервисных методов вебэлементы, уж лучше объект страницы на которой метод остановился. Иначе это очень странная логика получается.

UlyanaRomanova committed:
Исправление замечаний

DuskTraum commented:
- Не надо оборачивать GetRemoveSelector еще раз в try catch, у тебя уже в isElementPresent обернуто

UlyanaRomanova commented:
Я понимаю, но обернула в try catch, потому что без него он выдает исключение NoSuchElementException : no such element: Unable to locate element: {"method":"xpath","selector":"(//*[a='Coffee']/following-sibling::*[a='Remove']/a)[1]"}

DuskTraum commented:
> Я понимаю, но обернула в try catch, потому что без него он выдает исключение NoSuchElementException : no such element: Unable to locate element: {"method":"xpath","selector":"(//_[a='Coffee']/following-sibling::_[a='Remove']/a)[1]"}

Если правильно понимаю, это из-за того что ты сперва ищешь элемент         
private IWebElement linkRemoveTestProduct(string productName) => driver.FindElement(By.XPath($"(//*[a='{productName}']/following-sibling::*[a='Remove']/a)[1]"));
и только после этого идет проверка IsElementPresent.
Чтобы try catch сработал, то поиск элемента driver.FindElement должен быть внутри try catch
            try
            {
                return driver.FindElement(By.XPath($"(//*[a='{productName}']/following-sibling::*[a='Remove']/a)[1]"));
            }
По идее так должно сработать.

UlyanaRomanova committed:
Исправление замечания 2.

Убрала try catch из GetRemoveSelector и GetIdElementAfter. Поставила в начало при поиске элементов linkTestProduct и linkRemoveTestProduct.
Спасибо за подсказку.

DuskTraum commented:
> > Я понимаю, но обернула в try catch, потому что без него он выдает исключение NoSuchElementException : no such element: Unable to locate element: {"method":"xpath","selector":"(//_[a='Coffee']/following-sibling::_[a='Remove']/a)[1]"}
> 
> Если правильно понимаю, это из-за того что ты сперва ищешь элемент
> private IWebElement linkRemoveTestProduct(string productName) => driver.FindElement(By.XPath($"(//_[a='{productName}']/following-sibling::_[a='Remove']/a)[1]"));
> и только после этого идет проверка IsElementPresent.
> Чтобы try catch сработал, то поиск элемента driver.FindElement должен быть внутри try catch
> try
> {
> return driver.FindElement(By.XPath($"(//_[a='{productName}']/following-sibling::_[a='Remove']/a)[1]"));
> }
> По идее так должно сработать.

Сорян, может не очень ясно выразился, но я имел в виду, что ты можешь сделать метод IsElementPresent таким способом:
boolena IsElementPresent(By by) {
try {
driver.findElement(by)
}
catch (NoSuchElementException e)
{
return false;
}
return true;
}
Чтобы не писать такие здоровые методы как linkTestProduct, linkRemoveTestProduct

UlyanaRomanova committed:
Исправление. попытка № n

DuskTraum commented:
Принято
